src/Compilers/CSharp/Portable/Lowering/LocalRewriter/LocalRewriter_Call.cs

```diff
@@ -736,20 +753,44 @@ private static ImmutableArray<RefKind> GetRefKindsOrNull(ArrayBuilder<RefKind> r
                 }
                 else
                 {
+                    RefKind refKindForTemp = argRefKind;
+                    if (refKindForTemp == RefKind.In
+                        && !InArgumentRequiresCopy(rewrittenArguments[a], argumentsBinderOpt, discardedDiags))
+                    {
+                        refKindForTemp = RefKindExtensions.StrictIn;
+                    }
+
                     BoundAssignmentOperator assignment;
-                    var temp = _factory.StoreToTemp(argument, out assignment, refKind: argRefKind);
+                    var temp = _factory.StoreToTemp(
+                        argument,
+                        out assignment,
+                        refKind: refKindForTemp);
                     storesToTemps.Add(assignment);
                     arguments[p] = temp;
                 }
+
                 refKinds[p] = argRefKind;
             }
 
+            discardedDiags.Free();
+
             return;
 
             bool isLambdaConversion(BoundExpression expr)
                 => expr is BoundConversion conv && conv.ConversionKind == ConversionKind.AnonymousFunction;
         }
 
+        private static bool InArgumentRequiresCopy(
+            BoundExpression argumentExpr,
+            Binder argumentsBinder,
+            DiagnosticBag diagnostics)
```

Comment:
>DiagnosticBag diagnostics [](start = 12, length = 25)

It looks like all consumers discard diagnostics. Consider removing this parameter and discarding diagnostics locally. This might simplify consumers.